### ISSUE TITLE
Any plans to support multiple links bonding?

### ISSUE THREAD

**Issue by highfreq** [tailscale/tailscale#2256]
Wondering if tailscale supports or is planned to support in future any kind of link bonding.

thanks

**Comment by bradfitz** [comment]
Yes. In fact, I think I opened such a bug already in the past. (I have multiple ISPs and want to optimally use all of them)

**Comment by bradfitz** [comment]
Oh, the bug I'm thinking of is #1066. But I guess that's more about failover and not specifically about bonding. To confirm: is your interest in, say, splitting bandwidth or flows over multiple links? Or do you only care about failover? (That is, you'd primarily use your better link until it failed.)

**Comment by highfreq** [comment]
I am interested in bonding channels to add up bandwidth but using them as a failover too. Same as zerotier is trying to do (doesn't work at the moment).

**Comment by lukicweb** [comment]
This will be a so goo..

**Comment by felixn-unity** [comment]
@bradfitz 

I am not 100% on top of this, but I read an article with someone using wireguard and he had "bonding" enabled on the tunnel side, so multiple sessions/connections was going to the peer. Is that being implemented with TS?

Because I am looking at using TS as the main connection between our offices (we are using IPSEC now) - but if going that way, we need to get the maximum throughput)

Felix

**Comment by DentonGentry** [comment]
I do expect to support multiple wireguard connections taking multiple paths to increase throughput, but right now we only use one at a time.

Something that likely would work now is to create a bond0 interface from multiple eth# interfaces. That moves the bonding down to the Ip layer. This would require us to resolve https://github.com/tailscale/tailscale/issues/1066 and select the bond0 interface, but that might be more tractable.

**Comment by felixn-unity** [comment]
Ok so there is current only one ts peer connection? - most of our infrastrure is 10G and some offices/hubs have 10G internet.
 
So from what I understand this will not be utilized due to only one ts peer connection or? (so bond0 on eth# would not help I think)

Thanks
Felix

**Comment by DentonGentry** [comment]
bond0 exists to take multiple gigabit Ethernet interfaces like eth0, eth1, eth2, and eth3 and turn them into a bond0 interface which can carry an aggregate 4 Gbps of traffic.

For a native 10G Ethernet, there is no reason to use a bonding interface. There would be only a single link in the group, there would be no benefit.

**Comment by stp-ip** [comment]
Thought I share a project I ran into using Go in userspace to bond fully as in even fully aggregating multiple links for a single connection and not just multi session traffic.
https://github.com/benjojo/bondcat

**Comment by wizzard0** [comment]
I’d support this, but with a different flavor - optimizing not for bandwidth as such, but for latency/hiding packet loss (e.g. for live streaming), which together with an exit node would make for a cool self-hosted Speedify alternative (also, it’s just freaking cool)

**Comment by bradfitz** [comment]
@wizzard0, I plan to do exactly that for my home internet with Tailscale. It's on my list.

**Comment by scriptsrc** [comment]
I'm super curious how this would work in my environment. I have Starlink (unreliable, high throughput, high latency) and a local fixed wireless ISP (reliable-ish, slow 25mbps, amazingly low latency).

I previously used a dual-WAN USG in both load balancing mode and failover mode but the failovers were very slow and noticeable.

What would the characteristics be of bonding these with tailscale? 
- combined throughout?
- increased reliability?
- less noticeable "failovers"?
- what about latency?

**Comment by wizzard0** [comment]
@scriptsrc I routinely use Speedify to have peace of mind during video calls thru blackouts, bonding whatever available wired and wireless providers, be they at home, in the shelter or whatever.

Works wonders, no noticeable failovers over dozens of calls. Not a single time. More than compensates for a bit of added latency (since closest Speedify server is in Poland not Ukraine). But those guys are in the bonding business for years and it’s their core competence.

Still waiting for Tailscale though, because two VPNs manipulating route settings on the same laptop do go haywire sometimes.

**Comment by DentonGentry** [comment]
Because the original issue mentions throughput, Tailscale 1.36 substantially increased the throughput on Linux systems: https://tailscale.com/blog/throughput-improvements/

**Comment by DentonGentry** [comment]
Because the original issue mentions throughput, the 1.40 release contained additional throughput improvements described in [Surpassing 10Gb/s over Tailscale](https://tailscale.com/blog/more-throughput/)

**Comment by unixfox** [comment]
Does this feature request also cover the ability to automatically fallback to another link in case the main link is broken (let's say use wifi if ethernet do not work)? I'm not sure if tailscale already has this support or not. Or should I create a separate issue?

**Comment by OJFord** [comment]
@unixfox from https://github.com/tailscale/tailscale/issues/1066#issuecomment-942277657 I gather it's intended to support failover via #1552, though that's framed more as a one-time manual interface selection. (Or maybe it *is* this one, via 'if I bond a non-existent connection that path should have infinite cost'?)

**Comment by ArtemGr** [comment]
This is also known as https://en.wikipedia.org/wiki/Link_aggregation, no?

**Comment by dstnmrgn** [comment]
I recently started using speedify for streaming my son's baseball games. I would love to just be able to use Tailscale to do what I'm having to rely on speedify to do. Being able to bond multiple available WAN connections a device has when connecting to an exit node would be an awesome feature for Tailscale. Great for travel routers and for streaming sports events.

**Comment by dstnmrgn** [comment]
I asked about this on Ask a Tailscale Expert. https://www.youtube.com/live/o3WFaO0pAC0?t=599&si=TUgKrswFUlCk1Yi5

**Comment by thibaultmol** [comment]
@JayWStapleton ^Just a reminder that you brought this up, it would be really cool if we could have this in tailscale
